### DISCUSSION TITLE
Python Syntax Highlighting is incomplete compared to VSCode?

### DISCUSSION THREAD

**Discussion by Zeddicus414** [helix-editor/helix#2427] [Q&A] [answered]
Hi all,

Thanks for making Helix! It's really cool and I've been trying to set it up as my daily driver for python. One thing I've been trying to "fix" and I don't know where to look. Helix seems to be missing some distinctions in the syntax highlighting.

| Helix | VSCode |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/31261142/167266435-08b870aa-26e3-4936-a762-eec1ea091744.png) | ![image](https://user-images.githubusercontent.com/31261142/167266451-43103021-45c0-4541-afd2-9a1e0f1eecb4.png) |

In particular, notice the greenish types definitions on **Error**(**Exception**) and self.msg: **str**. Also, the purple control words, **if, else, return**

Is there some upstream library (maybe treesitter?) causing the lack of more detailed syntax highlighting found in VSCode?

I know the colors are configured in the dark_plus.toml theme file, so I don't think it's that:
![image](https://user-images.githubusercontent.com/31261142/167266572-c1da8386-2dea-43ca-885e-2e398f0c93d1.png)

Thanks!

**Comment by sudormrfbin** [comment]
(I edited the images in the question to be side by side so that it's easier to see the distinction if that's okay with you.)

The problem is that the highlight captures for python aren't as detailed as it could be, for example control keywords like `if`, `else`, are grouped under the `keyword` highlight group instead of the more specific `keyword.control` highlight group (which is assigned a different color in the dark plus theme):

https://github.com/helix-editor/helix/blob/ee889aaa854d0036da3bae16252bc382e50b0df6/runtime/queries/python/highlights.scm#L91-L121

As for the type names of `Error` and `Exception` they actually are different shade of blue than the highlighting of `class`, so I think that is more of a deviation with the theme definition of dark plus in helix.

For highlighting the `str` in `self.msg: str` properly we would need better tree sitter queries in the highlights file for python.

**Reply by Zeddicus414** [reply]
Ahh, just changing the highlights.scm to have a @keyword.control section makes a big difference. This is now from helix:
![image](https://user-images.githubusercontent.com/31261142/167268753-29882775-6fdd-4898-a733-aa4152900c51.png)

**Reply by Zeddicus414** [reply]
I am definitely in a place to start understanding this much better with a path forward for getting to where I want to be. Thank you!

**Comment by Zeddicus414** [comment, marked as answer]
Alright, this is where we are so far with help from  @sudormrfbin. See the attached highlights.scm file.

[highlights.txt](https://github.com/helix-editor/helix/files/8645916/highlights.txt)


| Helix | VSCode |
| --- | --- |
|![image](https://user-images.githubusercontent.com/31261142/167271550-553ceb20-3063-4796-b161-e5c68835432b.png)|![image](https://user-images.githubusercontent.com/31261142/167271558-693815cb-68c6-4ca2-a017-5999f0fec357.png)|


Some improvements:

- Types are green (line 6 & 25). This was done by placing the type definition more toward the top of the SCM file
- keyword.control allows most of the purple words that really increases readability, IMHO.

A few interesting things:

- I figured out that the class name (line 21) is being classified as a @constant, but I haven't figured out how to separate the class definition from an actual CONSTANT definition at the top of the code. In this case, I just made constants green too. I think it looks better. Notice the CONSTANT on line 3.
- VSCode makes imported modules the green color. I haven't found a way to duplicate that. See lines 1 & 19.
- VSCode is smart enough to tell when "in" is an operator vs part of the "for x in". See lines 8 & 9.

This is definitely close enough for my purposes. If I want to commit these changes to the projects, is that okay? I'm not sure if this can mess up any of the other themes. I think it's a clear improvement for this one. It's certainly good enough that I feel comfortable using it.

**Reply by archseer** [reply]
Feel free to open a PR!

**Reply by valpackett** [reply]
> tell when "in" is an operator vs part of the "for x in"

Quite easy to do here as well!

```lisp
(for_statement "in" @ui.linenr)
(for_in_clause "in" @ui.cursor.primary)
```

(place these *earlier* than the general "in" query)

![Screen_2022-05-08-19:21:22](https://user-images.githubusercontent.com/208340/167305383-149c0aa3-ad79-43e3-86b3-12ee72b12646.png)

(obviously use more sensible scopes, this is just for demonstration)

**Reply by Zeddicus414** [reply]
Ahh, that makes sense. Thank you!

I'm not particularly familiar with the treesitter scm syntax, but I'm getting there.

**Comment by qbedard** [comment]
Neovim's tree-sitter queries also seem better:
https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/python/highlights.scm

Is there a reason why we wouldn't want to use some unified set of queries between different editors? Are tree-sitter queries necessarily editor-specific?

**Reply by David-Else** [reply]
See the lightbulb: https://docs.helix-editor.com/master/guides/adding_languages.html?highlight=neovim#queries
